### ISSUE TITLE
Many schedulers seem to be heavily penalized by heavy I/O, at least with bcachefs

### ISSUE THREAD

**Issue by kode54** [sched-ext/scx#96]
I am using Arch, with linux-cachyos 6.7.0-4, with every upstream bcachefs patch from the 2024-01-01 tag up to 2bf0b0a9dff974cac259ce92d146e7142f472496 applied on top, with a bcachefs rootfs on a WD SN750 SSD, another bcachefs filesystem on a Samsung 980 Pro, and finally, my major storage on two 18TB WD Red Pro drives with 2 replicas enabled for both metadata and user data.

For heavy I/O, I am either running qBittorrent downloading some rather large (200GiB+) data sets at 1-2MB/s, plus some smaller ones (10-30GiB) at 20MB/s, to the two replicas hard drives. Or I am building a kernel on the rootfs with all 16 threads of my CPU.

Either of those tasks cause my compositor, Wayfire, to bog down heavily, if I happen to be using either scx_rustland, scx_rusty, or scx_nest. Disabling the sched-ext processes, and the compositor becomes performant again under the same conditions.

**Comment by arighi** [comment]
I'm wondering if this happens because we are giving too much priority to per-CPU kthreads, like during a massive I/O workload with fast storage drives, we may have per-CPU kernel workers that are actually CPU-bound more than IO-bound.

In rustland, for example, per-CPU kthreads are directly dispatched from the kernel to the local DSQ, bypassing the user-space scheduler. So they always win, over any other task and potentially they can up monopolizing the CPUs. Maybe something similar is happening also with the other schedulers?

Anyway, I was actually working on a patch to use the global DSQ for per-CPU kthreads (or better, for all the per-CPU threads in general). I'll do some tests and will create a new PR, I'll update this thread when it's ready.

**Comment by Decave** [comment]
TL;DR:

I think you're right that dispatching to `SCX_DSQ_GLOBAL` will actually in fact throttle the pcpu kthread, but some more relevant context below:

So there are a couple things to clarify here:

>IIUC from doc it seems that local DSQ wins over global DSQ.

Yes, that is correct. A core will first check its local DSQ to see if there are any tasks. Then it will check `SCX_DSQ_GLOBAL`, and only then will it call `ops.dispatch()`.

>So, let's say I have a single CPU, if I'm dispatching task1 with DSQ_LOCAL_ON | cpu and task2 with DSQ_GLOBAL, I would expect that task1 runs before task2.

So there is a concept in sched_ext called _direct dispatch_. This refers to when a task is dispatched "directly" from either `ops.select_cpu()` or `ops.enqueue()`, rather than waiting to be dispatched until `ops.dispatch()`. If you're doing direct dispatch, then you can't dispatch to `SCX_DSQ_LOCAL_ON | cpu` because we can't drop the rq lock on the enqueue path. You can dispatch directly to the local DSQ with `SCX_DSQ_LOCAL`, or you can dispatch directly to any other non-local DSQ.

So going back to your example, if you were to do a _direct_ dispatch of `task1` with `SCX_DSQ_LOCAL`, then it would indeed be chosen before `task2`. Otherwise, if you were to wait to dispatch `task1` until `ops.dispatch()`, and you instead did a direct dispatch of `task2`, `task2` will be chosen first because it will be consumed before `ops.dispatch()` is invoked, and thus before you can use `SCX_DSQ_LOCAL_ON | cpu`.

That said -- to tie all of that back to the example at hand -- you're correct that a task dispatched with `SCX_DSQ_LOCAL_ON | cpu` will take precedence over a task dispatched to `SCX_DSQ_GLOBAL`, so I actually do think what you're proposing should work to throttle the pcpu kthread. The only caveat potentially is that if tasks keep getting dispatched to that CPU from other CPUs (using `SCX_DSQ_LOCAL_ON` from `ops.dispatch()`, I think there's a possibility that the pcpu kthread could starve and actually never get to run, given that we'll always see that there's a task available on the local DSQ.

The crux of the issue is similar to what I alluded to above -- we're not actually getting to do things in `ops.dispatch()`. Perhaps the correct thing to do is to create per-cpu DSQs which we dispatch everybody to instead of using `SCX_DSQ_LOCAL_ON`? That would look something like this:
- In `rustland_init()`, we create a custom DSQ per CPU
- In `rustland_enqueue()`, rather than dispatching to `SCX_DSQ_LOCAL` for pcpu kthreads, we dispatch to `RUSTLAND_DSQ_N` (`N` == cpu where task is being enqueued).
- In `rustland_dispatch()`, rather than dispatching a task to `SCX_DSQ_LOCAL_ON | task.cpu`, we instead dispatch to `RUSTLAND_DSQ_N` where `N == task.cpu`. After dispatching, we consume all of the tasks from `RUSTLAND_DSQ_N` with `scx_bpf_consume()`.

That would allow us to always call `dispatch_user_scheduler()` from `rustland_dispatch`, while also giving us the same FIFO semantics we're already encountering with using `SCX_DSQ_LOCAL{_ON}`. In other words, we'll be calling `rustland_dispatch()` a bit more, but it should be minimal and also avoid the issue you pointed out. Let me try putting together a PR for this.

**Comment by Decave** [comment]
Here's an example of what I was talking about: https://github.com/sched-ext/scx/commit/f4a7cb2e0903ddca3884fe819d023844b032ca2b. It doesn't perform very well though :-(

**Comment by arighi** [comment]
@Decave thank you so much for all the details and the example! It's all clear now.

About the tasks dispatched to the global DSQ, yes we may have starvation. And I like your idea about using a DSQ per-cpu. About the poor performance, how about kicking the cpu when we're sending task to a different one?

Something like this:
```
--- a/scheds/rust/scx_rustland/src/bpf/main.bpf.c
+++ b/scheds/rust/scx_rustland/src/bpf/main.bpf.c
@@ -515,6 +515,9 @@ void BPF_STRUCT_OPS(rustland_dispatch, s32 cpu, struct task_struct *prev)
                dbg_msg("usersched: pid=%d cpu=%d payload=%llu",
                        task.pid, task.cpu, task.payload);
                dispatch_task(p, task.cpu, 0);
+               if (cpu != task.cpu)
+                       scx_bpf_kick_cpu(task.cpu, 0);
+
                __sync_fetch_and_add(&nr_user_dispatches, 1);
                bpf_task_release(p);
```

**Comment by Decave** [comment]
Well, that makes the scheduler work great!

**Comment by arighi** [comment]
@Decave I've done some some quick tests and with the extra kick the scheduler doesn't seem bad at all. I'll do more tests tomorrow morning, but I think we may have a solution, at least for my other cpumask/affinity issue, `stress-ng --race-sched N` doesn't crash the scheduler now.

**Comment by kode54** [comment]
I have tried out #99 against my system, and it does seem to work for a while, but then randomly devolves into horrible stuttering, both moving my mouse cursor and moving windows around the desktop.

**Comment by arighi** [comment]
> I have tried out #99 against my system, and it does seem to work for a while, but then randomly devolves into horrible stuttering, both moving my mouse cursor and moving windows around the desktop.

Sorry @kode54 , #99 doesn't address yet any IO pressure issue, but it was required for a follow up patch that I'm planning to test (the idea is to send all kthreads to the user-space scheduler as well as the regular tasks), that maybe can help to handle the situation. But thanks for testing this one and good to know it helped a bit with this issue. I'll keep you informed when I have something ready for testing.

**Comment by arighi** [comment]
@kode54 can you test #99 now? I added more changes that seem to help under IO pressure condition (as a stress test I'm running a bunch of `fio` writers on an encrypted dm-crypt volume). I'm not sure that I'm reproducing exactly your problem, but in my case the system seems more responsive now. Thanks!

**Comment by kode54** [comment]
Still stuttering, but in this case, the load is OBS Studio running a virtual camera and using 60-80% of a core for that, and Discord video chat using the same amount on another core, and my compositor using a steady 10% of a core or so, or slightly more if I start moving windows around.

**Comment by arighi** [comment]
> Still stuttering, but in this case, the load is OBS Studio running a virtual camera and using 60-80% of a core for that, and Discord video chat using the same amount on another core, and my compositor using a steady 10% of a core or so, or slightly more if I start moving windows around.

Hm.. ok, so in this case you don't even have any heavy IO-bound task. I'm curious, do you have SMT enabled? If possible can you try to add `nosmt` to your grub config, reboot and see if it's better (or worse)?

**Comment by kode54** [comment]
SMT is enabled, disabling it did not have any appreciable effect, other than reducing me to 8 threads. Still stuttering while under CPU load.

On login, Steam downloads and compiles shaders for about 20 different installed games. It seems to do this about every day.

**Comment by arighi** [comment]
> SMT is enabled, disabling it did not have any appreciable effect, other than reducing me to 8 threads. Still stuttering while under CPU load.
> 
> On login, Steam downloads and compiles shaders for about 20 different installed games. It seems to do this about every day.

OK, good! At least we know it's not SMT-related. With the same workload the stuttering happens also with other schedulers, like scx_simple, scx_rusty, etc. or it's just scx_rustland?

**Comment by kode54** [comment]
Under certain loads, I experienced stutter with scx_rusty and scx_nest. I can try scx_simple as well.

Does it matter that the kernel I am using the patch set with is linux-cachyos? Built using the default settings, except for using generic cpu instead of autodetect, and adding a bcachefs update patch set that shouldn’t be incurring a significant load versus plain 6.7.1.

**Comment by arighi** [comment]
It shouldn't matter, if the default scheduler works fine, then the sched-ext schedulers should also work fine, to a large degree, meaning not for everything, but they shouldn't show obvious lags or stuttering, especially with workloads that are not super intense.

But even if scx_simple clearly shows this problem, then maybe something odd is happening in the kernel.

**Comment by kode54** [comment]
scx_simple seems to be fine, so far.

**Comment by arighi** [comment]
> scx_simple seems to be fine, so far.

ok, in this case it seems reasonable to assume that this issue has nothing to do with the kernel or sched-ext in general.

So, I would suggest another test: can you try #99 again (I pushed/updated stuff, so make sure you refresh the repo) and start scx_rustland with `-b 0`? This disables the priority boost logic for interactive tasks and the scheduler becomes pretty much a vruntime-based scheduler with a variable time slice, that is very similar to what scx_simple does.

This would tell us the problem (for rustland at least) is in the priority boot part or somewhere else.

**Comment by arighi** [comment]
@kode54 I think I was able to reproduce the problem on my side, starting obs and recording my session. I've updated #99 again, adding a WIP patch (make sure to git reset the repo). I'm still not happy about this patch (hence the WIP), but it seems to make a significant difference in my case and it might fix the problem also in your case.

Can you do another test? Thanks.

**Comment by kode54** [comment]
I'll test shortly. Sorry if I wasn't much help the first post, I wasn't up yet.

**Comment by kode54** [comment]
Was running your latest #99 of scx_rustland, and it was working mostly fine, until I attempted to rebuild the scx-scheds package again, which with my AUR defaults, attempts over 30-40 threads of rustc in the final stage. This led to my GPU timing out and resetting, and the reset took so long, my desktop session crashed back to login, terminating the build.

**Comment by arighi** [comment]
hm... it should survive to a 30-40 threads build, I'll try to run some parallel rustc builds also on my box.

**Comment by kode54** [comment]
I will leave this issue open, but I'm ceasing testing for now, until sched-ext makes it into the upstream kernel tree in a stable release. I won't be testing any 6.8 kernels until Arch has a 6.8 kernel in the `linux` package, and I will only be using official binaries for this, since I can't otherwise report a userland package being incompatible.

**Comment by kode54** [comment]
Okay, I've returned to continue testing this, because I overcame my stupid problem. I will simply live within the issues of rolling new kernels.

Anyway:

I encountered further bugs with scx_rustland, but not with scx_rusty or other BPF schedulers that aren't entirely userspace. Basically, if I have the `scx-scheds-git` AUR PKGBUILD, and a makepkg.conf set to use `-j$(nproc)` on my 16 thread CPU, it will queue up all four of the rust build jobs at once, and each one of them will use up to 16 threads automatically without regard for each other.

This build queue has immediate problems with my Radeon RX 6700-XT on Wayfire, but only with scx_rustland. It will almost immediately result in a GPU reset, which will end up failing, leaving the GPU broken until the machine is soft rebooted using SSH to log in remotely.

scx_rusty, the default in `/etc/default/scx` shipped with the current master, survives the build process. It is slightly stuttery, but so is the kernel built-in scheduler.

Here is a dmesg dump of the failing scx_rustland session:
[dmesg.1.txt](https://github.com/sched-ext/scx/files/14079412/dmesg.1.txt)

And here is a dmesg dump of the successful session which followed, running on scx_rusty:
[dmesg.2.txt](https://github.com/sched-ext/scx/files/14079415/dmesg.2.txt)

**Comment by arighi** [comment]
@kode54 thank you for sharing this, it's very useful to understand what's happening! It looks like you hit a sched-ext issue, more than a rustland issue, more exactly this warning:

```
int scx_cgroup_can_attach(struct cgroup_taskset *tset)
{
...
                WARN_ON_ONCE(p->scx.cgrp_moving_from);
...
```

I'm wondering if we need to exclude exiting / autogroup tasks in this logic, something similar to what we did in https://github.com/sched-ext/sched_ext/commit/6b747e0ee5fca284330d065a0c777d1991290bc4.

I'm not sure if the following patch makes any sense at all (posting here just for discussion):
https://github.com/arighi/sched_ext/commit/6f51182258fa6ff9e90393ba2bd73fc6e8f30358

Maybe if you have the time / possibility to recompile the kernel and do more tests you can check if the problem is still happening with this patch applied. Otherwise, let's wait for the opinion of more experienced people than me in this topic, such as @htejun or @Decave .

**Comment by kode54** [comment]
It still reset my GPU with that patch applied. scx_rusty did not, once again.

scx_rustland: [dmesg.3.txt](https://github.com/sched-ext/scx/files/14081703/dmesg.3.txt)
scx_rusty: [dmesg.4.txt](https://github.com/sched-ext/scx/files/14081722/dmesg.4.txt)

**Comment by arighi** [comment]
oh wow, that was fast, thanks!

Alright, it looks like it has nothing to do exiting tasks / autogroup, then maybe let's wait for a feedback from the cgroup experts, so that we can do better focused tests.

**Comment by kode54** [comment]
This may be relevant as to why my gpu resets so easily. I had a gpu issue with a different thing at some point, and mentioned this in the arch linux main irc channel, and someone posted a snippet of kernel command line arguments, which I applied. Among them is an option that sets the amdgpu timeout to one second. Here’s the whole lot:

```
amdgpu.runpm=0 amdgpu.aspm=0 amdgpu.bapm=0 amdgpu.audio=0 amdgpu.gpu_recovery=1 amdgpu.noretry=0 amdgpu.lockup_timeout=1000
```

**Comment by arighi** [comment]
Anything is possible... usually if the scheduler does something bad you may experience stalls / starvation. A GPU reset or wayland crashing, seem more like an issue not related to the scheduler. It could be triggered by the scheduler, meaning that a certain scheduler can enforce the exact timing constraints so that a certain problem happens. 

In the latest dmesg posted by @kode54 we can clearly see that a kernel bug is happening (cgroup / sched-ext related). @ptr1337 do you see the same errors in your dmesg?

**Comment by kode54** [comment]
I solved the GPU resets on building the AUR package by removing that snippet from my kernel command line. The GPU still stalls briefly once, but recovers within the default 10s timeout that is the default, not the 1s that was specified by that snippet.

There were some other problems, including rustland timing out under such an insane load and being restarted repeatedly, but nothing earth shattering as far as rendering the system unusable without a reboot:

[dmesg.5.txt](https://github.com/sched-ext/scx/files/14121267/dmesg.5.txt)

**Comment by kode54** [comment]
scx_rustland seems a lot better now, and handles an insane workload just fine. Though I am no longer using bcachefs.

**Comment by arighi** [comment]
@kode54 thanks for the feedback! BW what was your usual workload with bcachefs? I'd like to try to reproduce it and see if I hit any issue with the latest rustland. Was it like large stream of read/writes? Or more like short scattered I/O? Thanks.

**Comment by kode54** [comment]
Not sure if it’s the same with btrfs, but it was a raid1 type setup on two 7200 rpm drives, running a bunch of large (10+GB each) torrents with qbittorrent. Mostly downloading, with a top downstream hit of about 20-40MB/s, and maybe 2MB/s up. Maybe mass download some Linux isos for testing.

**Comment by kode54** [comment]
Found a new bottleneck with rustland: Fall Guys running under Proton. It has heavy lag unless I stop rustland. I attempted to test rusty, but it did the same thing, and eventually froze my entire machine.

**Comment by kode54** [comment]
When I run with rustland, it doesn't freeze the system. I will need to test rusty again. No errors are logged, though I didn't check the scx journal, I'll check it if I can reproduce it again.

When I stop the scx service / rustland, Fall Guys client stays about the same, between 70% and 100% or slightly more, depending on what's rendering, but when it's stopped, there's no major stuttering. When I start it, the wineserver process usage goes from 30ish % to 50-60%, and scx_rustland is using about 27%.

I'll now try to measure with scx_rusty and see if I can reproduce the freeze.

**Comment by arighi** [comment]
@kode54 about rustland I pushed some improvements yesterday that should reduce the cpu usage and improve responsiveness in general. Do you mind trying again with the latest version from the `main` branch and see if the stuttering is still the same / better / worse? Thanks!

**Comment by kode54** [comment]
I tried with both rusty and rustland, stuttering is pretty bad, and the wineserver process is using 20-30% more CPU according to `top`, compared to the kernel scheduler. I can't reproduce the lockup, but maybe that was because I started rusty while the game was already running.

I don't know if it's worth noting that I was using the Steam version of the game, which is no longer available for purchase. I can try the Epic Games Store version under Heroic Games Launcher, if you think that will help.

**Comment by Decave** [comment]
`rusty` still has a lot of room for improvement -- it's mostly been targeted towards server workloads so far. My plan is to start looking at making it more interactive in the near future.

**Comment by arighi** [comment]
Hm.. I've tried a bunch of games with rustland and I can't reproduce any stuttering, fps is always really close to the default scheduler (or even better is the system is busy).

@kode54 maybe you can try to run `perf top` when the stuttering is happening, that should help to identify the bottleneck.

Even better, you could try to generate a more detailed profiling using this command (maybe run it for like 30 sec or similar when the stuttering is happening, the ctrl+c and post the output):
```
sudo bpftrace -e 'profile:hz:99 { @[ustack, kstack] = count(); }'
```

This would tell us where the system is spending most of the time, showing both the kernel and the user stack trace of all the running processes.

**Comment by kode54** [comment]
I have generated a lengthy log of tracing from roughly before launching the game, to launching it, which stutters on loading.

It also appears my machine is using 7GiB of ZSWAP, ZSTD compressed. Even though it is only using 8 GiB of application memory, it's not relinquishing much of its 21 GiB of cached file data.

Could you tell me how I should analyze this data, if I want to inspect it myself?

[bpftrace.txt](https://github.com/sched-ext/scx/files/14372471/bpftrace.txt)

**Comment by htejun** [comment]
Also, as you mentioned swap, can you record the output of `/proc/pressure/memory` every 10 or so seconds and see what it says?

**Comment by kode54** [comment]
Here's a memory pressure log. The increase in 10 second average levels was right around where the game was loading. I do not know why it is swapping so much when I have 32GB of RAM.
[memorylog.txt](https://github.com/sched-ext/scx/files/14383893/memorylog.txt)

**Comment by htejun** [comment]
It's not necessarily swapping. If it mmapped large memory areas and faulting them in for loading, they'd show up as memory pressure, which isn't too surprising while loading. Just so that we can rule out memory/io issues, you still see stuttering problems after the pressure spike from loading subsides?

**Comment by htejun** [comment]
So, the stuttering problems are associate with memory pressure? If so, I wonder whether this is from the schedulers always dispatching per-cpu kworkers directly to local DSQ prioritizing them over everything else.

**Comment by kode54** [comment]
`nr_page_faults` never exceeds 0. Though when I launched scx_rustland, `tasks` was about 60-70, then increased to about 700 by the time I had the game running. Also, most of the cores were listing `pid=0` most of the time, except when 4 processes from other PIDs cycled around the various cores every second of output.

**Comment by arighi** [comment]
ok `nr_page_faults=0` is good, 700 waiting tasks, instead, is not really good, are they all listed in `nr_queued`, in `nr_scheduled` or both? If they're all in `nr_queued` it means that the scheduler is not awakened fast enough, if they're in `nr_scheduled` it means that the scheduler should be more aggressive at dispatching them.

Listing `pid=0` in most of the cores is normal (unless the system is massively overloaded), because the scheduler runs when some tasks expire their time slice and in order to be able to dispatch more tasks at least some cores needs to be free.

**Comment by kode54** [comment]
Here's a sample of output with a queued task:

```
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO] vruntime=185057608512
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   tasks=988
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   nr_user_dispatches=3077278 nr_kernel_dispatches=57754827
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   nr_cancel_dispatches=0 nr_bounce_dispatches=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   nr_waiting=1 [nr_queued=1 + nr_scheduled=0]
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   nr_failed_dispatches=0 nr_sched_congested=0 nr_page_faults=0 [OK]
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO] time slice = 20000 us
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO] slice boost = 1600
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO] Running tasks:
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  0 cpu  0 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  0 cpu  8 pid=184230
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  1 cpu  1 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  1 cpu  9 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  2 cpu  2 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  2 cpu 10 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  3 cpu  3 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  3 cpu 11 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  4 cpu  4 pid=686362
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  4 cpu 12 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  5 cpu  5 pid=1289884
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  5 cpu 13 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  6 cpu  6 pid=[self]
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  6 cpu 14 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  7 cpu  7 pid=0
Feb 27 02:05:53 mrgency bash[1286191]: 10:05:53 [INFO]   core  7 cpu 15 pid=0
```

**Comment by arighi** [comment]
That looks correct, things can be problematic if `nr_queued` or `nr_scheduled` become big numbers.

**Comment by kode54** [comment]
I attempted to run it with `PROTON_LOG=1` to see if there would be anything useful. Just 375MB of log spew over 100 seconds. Most of it is exceptions being thrown and stack unwind being logged, multiple times per *millisecond*.

**Comment by kode54** [comment]
I doubled my RAM capacity, and it's still happening. No swapping going on.

I checked Netdata. Average CPU Pressure hits about 17% at its peak while the game is loading, then drops to about 10% while it's running idle. CPU utilization hits 38% or so while it's loading, and that's across all 16 threads.

**Comment by kode54** [comment]
Further checking: I'm compiling a kernel now, while a btrfs scrub is running on my large storage array, which isn't involved in the compilation or its backing. I was hitting 98-100% CPU usage and load averages of 19 or so for my 16 threads.

Then I loaded scx_rustland. CPU usage dropped to 78% and load averages shot up to 30.

**Comment by kode54** [comment]
It seems to be working fine so far with `-s 5000` on `scx_rustland`. I’m not sure if previous updates from 6.7.7 affecting zswap had any effect on this. Also the game seems to use more cpu time if I have vsync enabled. I’m also using variable refresh rate, 40-60Hz.

**Comment by arighi** [comment]
Sometimes I also need to use `-s 5000` to prevent some audio cracking when the system is massively loaded. Or (not sure why) I need to change the scheduling class of pipewire/pipewire-pulse/wireplumber from real-time to normal (so that they'll be scheduled by rustland as well). Either way allows me to prevent the audio cracking.

I'm really considering to set the default time slice to 5000 (or 10000) for `scx_rustland`, considering that the main goal of this scheduler is to prioritize low-latency workloads, probably it makes sense to use a shorter time slice by default.

**Comment by arighi** [comment]
I think we can close this one for now, after the latest changes to the default settings in scx_rustland we should be able to mitigate these performance issues.

@kode54 do you agree or do you think there's something else that we should address/improve?

**Comment by kode54** [comment]
scx_rusty needs to be 5ms too, unless low latency isn’t its aim.

**Comment by arighi** [comment]
scx_rusty is a more general purpose scheduler, so it also needs to take throughput into account. Do you see some improvements in your case If you run `scx_rusty -s 5000`?

I know @Byte-Lab had some plans to apply the same "dynamic time slice" concept also to scx_rusty, that should help in cases like this. I'm not sure if there's some work in progress for this, otherwise I can take a look.

So, let's keep this open for now. Thanks @kode54 .

**Comment by kode54** [comment]
I’m not sure if it’s kernel updates or using `-s 5000`, but I can now play Fall Guys without terrible lag. Also, unlike rustland, rusty doesn’t impact my Geekbench score by over 3000 points on the multi core result. My multi core score is now more than the average score for this model CPU.

**Comment by Byte-Lab** [comment]
Nice, glad to hear. We still have a lot more we can do to make rusty  more interactive, but glad to hear things seem to be going in the right direction.

**Comment by arighi** [comment]
> I’m not sure if it’s kernel updates or using `-s 5000`, but I can now play Fall Guys without terrible lag. Also, unlike rustland, rusty doesn’t impact my Geekbench score by over 3000 points on the multi core result. My multi core score is now more than the average score for this model CPU.

About rustland this is kind of expected, the scheduler is not really nice with non-interactive cpu-intensive tasks. :smile:

**Comment by kode54** [comment]
Adding another weird case: Running an OBS Studio virtual camera with v4l2loopback, feeding it a combined PipeWire desktop capture and webcam capture overlay, and it becomes incredibly laggy when the CPU is fully loaded, such as from building a package or a kernel.

**Comment by arighi** [comment]
I think we've made quite good progress about the issues reported here, we can probably close this at this point, @kode54 anything you want to add / are you still experiencing any issue in particular? Thanks.

**Comment by kode54** [comment]
I can’t add more for now anyway, as I’m stuck on an LTS kernel until this bug is dealt with:

https://gitlab.freedesktop.org/drm/amd/-/issues/3142
